src/http/curl.h

```diff
@@ -46,25 +62,598 @@ namespace ccf::curl
+        throw std::logic_error(
+          "Cannot attach request body to a null CURL handle");
+      }
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_READDATA, this);
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_READFUNCTION, send_data);
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_INFILESIZE, unsent.size());
+    }
+  };
+
+  class Response
+  {
+  public:
+    std::vector<uint8_t> buffer;
+    using HeaderMap = std::unordered_map<std::string, std::string>;
+    HeaderMap headers;
+    long status_code = 0;
+
+    static size_t write_response_chunk(
+      uint8_t* ptr, size_t size, size_t nmemb, Response* response)
+    {
+      if (response == nullptr)
+      {
+        LOG_FAIL_FMT(
+          "write_response_chunk called with a null response pointer");
+        return 0;
+      }
+      auto bytes_to_copy = size * nmemb;
+      response->buffer.insert(response->buffer.end(), ptr, ptr + bytes_to_copy);
+      // Should probably set a maximum response size here
+      return bytes_to_copy;
+    }
+
+    static size_t recv_header_line(
+      char* buffer, size_t size, size_t nitems, Response* response)
+    {
+      if (response == nullptr)
+      {
+        LOG_FAIL_FMT("recv_header_line called with a null response pointer");
+        return 0;
+      }
+      auto bytes_to_read = size * nitems;
+      std::string_view header(buffer, bytes_to_read);
+
+      // strip /r/n etc
+      header = ccf::nonstd::trim(header);
+
+      // Ignore empty headers, and the http response line (e.g. "HTTP/1.1 200")
+      static const std::regex http_status_line_regex(R"(^HTTP\/[1-9]+.*)");
+      if (
+        !header.empty() &&
+        !std::regex_match(std::string(header), http_status_line_regex))
+      {
+        const auto [field, value] = ccf::nonstd::split_1(header, ": ");
+        if (!value.empty())
+        {
+          std::string field_str(field);
+          nonstd::to_lower(field_str);
+          response->headers[field_str] = ccf::nonstd::trim(value);
+        }
+        else
+        {
+          LOG_INFO_FMT("Ignoring invalid-looking HTTP Header '{}'", header);
+        }
+      }
+
+      return bytes_to_read;
+    }
+
+    void attach_to_curl(CURL* curl)
+    {
+      if (curl == nullptr)
+      {
+        throw std::logic_error("Cannot attach response to a null CURL handle");
+      }
+      // Body
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_WRITEDATA, this);
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_WRITEFUNCTION, write_response_chunk);
+      // Headers
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_HEADERDATA, this);
+      CHECK_CURL_EASY_SETOPT(curl, CURLOPT_HEADERFUNCTION, recv_header_line);
+    }
+  };
+
+  class CurlRequest
+  {
+  private:
+    UniqueCURL curl_handle;
+    std::string url;
+    ccf::curl::UniqueSlist headers;
+    std::unique_ptr<ccf::curl::RequestBody> request_body = nullptr;
+    std::unique_ptr<ccf::curl::Response> response = nullptr;
+    std::optional<std::function<void(CurlRequest&)>> response_callback =
```

Comment:
I suggest add pretty name for types like this, as soon as we have to spell it twice (here as both a field and constructor arg).

`using ResponseCallback = std::function<void(CurlRequest&)>;` at the class or namespace scope.